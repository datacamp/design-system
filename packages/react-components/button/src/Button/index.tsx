import * as Icons from '@datacamp/waffles-icons';
import { Text } from '@datacamp/waffles-text';
import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import { computeDataAttributes } from '@datacamp/waffles-utils';
import { css, SerializedStyles } from '@emotion/core';
import { childrenOfType, nChildren } from 'airbnb-prop-types';
import PropTypes from 'prop-types';
import React, { ComponentProps, ReactElement, useRef, useState } from 'react';
import { useUIDSeed } from 'react-uid';
import { mergeRefs } from 'use-callback-ref';

import Tooltip from '../Tooltip';
import {
  baseColors,
  baseLoadingStyle,
  baseStyle,
  fontSizes,
  getAppearanceStyle,
  getDisabledStyle,
  getIconSize,
  getSize,
} from './buttonStyles';
import Spinner from './spinner';

type IconType = typeof Icons.AddCircleIcon; // Could use any Icon here
type IconElement = ReactElement<ComponentProps<IconType>, IconType>;

interface BaseButtonProps {
  /**
   * Sets the css class of the rendered element. Can be used to apply custom
   * styles.
   */
  className?: string;
  /**
   * As with all the other waffles components, dataAttributes can be used to set
   * data- html attributes on the element.
   */
  dataAttributes?: { [key: string]: string };
  /**
   * The button can be disabled by passing "disabled" as a prop.
   */
  disabled?: boolean;
  /**
   * By passing "loading" as a prop, the button will be disabled and it will
   * render a spinner.
   */
  loading?: boolean;
  /**
   * The size of the button to render.
   */
  size?: 'small' | 'medium' | 'large';
  /**
   * When tooltipText is provided, this sets the colour of the tooltip. It
   * should be used make the tooltip visible on different colour backgrounds.
   */
  tooltipAppearance?: 'light' | 'dark';
  /**
   * When tooltipText is provided, this determines the position of that text.
   * "top", "bottom", "left", and "right" will position the tooltip at the
   * centre of that side. The other styles can be used to prevent the tooltip
   * from overflowing when it is longer than the button.
   */
  tooltipPosition?:
    | 'bottom'
    | 'top'
    | 'left'
    | 'right'
    | 'bottomRight'
    | 'bottomLeft'
    | 'topLeft'
    | 'topRight';
  /**
   * An additional description to show on hover or focus.
   */
  tooltipText?: string;
}

interface LinkButtonProps {
  href: string;
  target?: '_blank' | undefined;
  type: 'link';
}

interface ButtonButtonProps {
  onClick: () => void;
  type?: 'button' | undefined;
}

interface SubmitButtonProps {
  type: 'submit';
}

interface StringChildProps {
  ariaLabel?: string;
  children: string;
}

interface IconChildProps {
  ariaLabel: string;
  /**
   * When providing a react element as a child, it can only be one of the Icon
   * components exposed by @datacamp/waffles-icons
   */
  children: IconElement;
}

interface DefaultProps {
  appearance?: 'default';
  intent?: intents;
}

interface PrimaryProps {
  appearance: 'primary';
  intent?: intents | 'cta';
}

type intents = 'neutral' | 'danger' | 'success' | 'warning';

interface IconTextChildProps {
  ariaLabel?: string;
  children: [string, IconElement] | [IconElement, string];
}

export type ButtonProps = BaseButtonProps &
  (ButtonButtonProps | LinkButtonProps | SubmitButtonProps) &
  (StringChildProps | IconChildProps | IconTextChildProps) &
  (DefaultProps | PrimaryProps);

const InternalButton = (
  props: ButtonProps & { innerRef?: React.Ref<any> },
): React.ReactElement => {
  const {
    appearance = 'default',
    ariaLabel,
    children,
    className,
    dataAttributes,
    disabled = false,
    innerRef,
    intent = 'neutral',
    loading = false,
    size = 'medium',
    tooltipAppearance = 'dark',
    tooltipText,
    tooltipPosition = 'bottom',
  } = props;

  const buttonRef = useRef<HTMLElement>();
  const [hasFocus, setHasFocus] = useState(false);
  const [hasHover, setHasHover] = useState(false);
  const uidSeed = useUIDSeed();
  const tooltipId = uidSeed('button-tooltip');
  const outlineTextColor = tokens.color.opaque.greyDark.value.hex;
  const ctaTextColor = intent === 'cta' ? outlineTextColor : 'white';
  const outlineIconColor = disabled ? baseColors[intent] : 'currentColor';
  const textColor = appearance === 'primary' ? ctaTextColor : outlineTextColor;
  const iconColor = appearance === 'primary' ? ctaTextColor : outlineIconColor;

  const baseTextStyle = css(fontSizes[size], {
    color: loading ? 'transparent' : textColor,
    fontWeight: 'bold',
  });
  const margin =
    size === 'large' ? tokens.size.space[16].value : tokens.size.space[8].value;
  const getTextStyleWithMargin = (i: number | undefined): SerializedStyles => {
    if (i === 1) {
      return css(baseTextStyle, {
        marginLeft: margin,
      });
    }
    return css(baseTextStyle, {
      marginRight: margin,
    });
  };
  const parsedDataAttributes = computeDataAttributes(dataAttributes);
  const buttonStyle = css(
    baseStyle,
    getAppearanceStyle(appearance, intent, !loading && !disabled),
    disabled && getDisabledStyle(appearance, intent),
    loading && baseLoadingStyle,
    React.Children.count(children) > 1 || typeof children === 'string'
      ? getSize(size)
      : getIconSize(size),
  );

  // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
  const { Element, ...otherProps } = (() => {
    switch (props.type) {
      case 'link':
        return {
          Element: 'a',
          href: props.href,
          target: props.target,
        } as const;
      case 'submit':
        return { Element: 'button', type: 'submit' } as const;
      case 'button':
      default:
        return {
          Element: 'button',
          onClick: !loading ? props.onClick : undefined,
          type: 'button',
        } as const;
    }
  })();

  const tooltipVisible = !(disabled || loading) && (hasFocus || hasHover);

  return (
    <>
      <Element
        ref={innerRef ? mergeRefs([buttonRef, innerRef]) : buttonRef}
        aria-describedby={tooltipText && tooltipVisible ? tooltipId : undefined}
        aria-label={ariaLabel}
        className={className}
        css={buttonStyle}
        disabled={disabled || loading}
        onBlur={() => setHasFocus(false)}
        onFocus={() => setHasFocus(true)}
        onMouseEnter={() => setHasHover(true)}
        onMouseLeave={() => setHasHover(false)}
        {...otherProps}
        {...parsedDataAttributes}
      >
        {loading && (
          <Spinner
            css={{ position: 'absolute' }}
            inverted={appearance === 'primary' && intent !== 'cta'}
          />
        )}
        {React.Children.map(children, (child, i) =>
          typeof child === 'string' ? (
            <Text
              css={
                React.Children.count(children) > 1
                  ? getTextStyleWithMargin(i)
                  : baseTextStyle
              }
            >
              {child}
            </Text>
          ) : (
            React.cloneElement(child, {
              'aria-hidden': true, // hide icon from screen reader so only ariaLabel or button text is read.
              color: loading ? 'transparent' : child.props.color ?? iconColor,
              size: size === 'large' ? 24 : 18,
              title: '', // remove tooltip from icon within button
            })
          ),
        )}
      </Element>
      {tooltipText && (
        <Tooltip
          appearance={tooltipAppearance}
          id={tooltipId}
          position={tooltipPosition}
          target={buttonRef}
          visible={tooltipVisible}
        >
          {tooltipText}
        </Tooltip>
      )}
    </>
  );
};

// additional prop-types validation that the child is either icon or string
const iconValidator = childrenOfType(...Object.values(Icons));

InternalButton.propTypes = {
  /**
   * When the appearance is "primary", the button will have a filled colour.
   */
  appearance: PropTypes.oneOf(['default', 'primary']),
  /**
   * Set the aria-label on the rendered element. This is required when using a single icon as a child.
   */
  ariaLabel: PropTypes.string,
  /**
   * The content of the button. This can either be text, and icon, or a
   * combination of the two. When this is one of the components exposed by
   * @datacamp/waffles-icons, the button will render as a square. In this
   * situation ariaLabel is required.
   */
  children: PropTypes.oneOfType([
    nChildren(
      2,
      PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, iconValidator])),
    ),
    PropTypes.string,
    iconValidator,
  ]).isRequired,
  /**
   * The destination of the link. Only available when type="link".
   */
  href: PropTypes.string,
  /**
   * Defines the color of the button. The cta intent can be used only with
   * the primary appearance.
   */
  intent: PropTypes.oneOf(['neutral', 'danger', 'success', 'warning', 'cta']),
  /**
   * The callback fired when the button is clicked. Only available when
   * type="button".
   */
  onClick: PropTypes.func,
  /**
   * Specifies where to open the linked document. Only available when
   * type="link".
   */
  target: PropTypes.string,
  /**
   * Determines what is rendered, and what other props are available. The type
   * "submit" should be used for form submissions. The type "link" will render
   * an <a> tag, and should be used with an 'href'. And the type "button" should
   * be used with the 'onClick' prop.
   */
  type: PropTypes.oneOf(['button', 'link', 'submit']),
};

InternalButton.defaultProps = {
  appearance: 'default',
  disabled: false,
  intent: 'neutral',
  loading: false,
  size: 'medium',
  tooltipAppearance: 'dark',
  tooltipPosition: 'bottom',
  type: 'button',
};

const Button = React.forwardRef<any, ButtonProps>((props, ref) => (
  <InternalButton innerRef={ref} {...props} />
));

export default Button;
