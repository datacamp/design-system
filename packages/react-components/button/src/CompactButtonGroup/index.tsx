import { ClassNames } from '@emotion/core';
import { childrenOfType } from 'airbnb-prop-types';
import PropTypes from 'prop-types';
import React from 'react';

import Button, { ButtonProps } from '../Button';

interface CompactButtonGroupProps {
  /**
   * The content of the CompactButtonGroup to render. This must be a set of
   * Button components. All Buttons within a CompactButtonGroup must have the
   * same appearance, intent, size & disabled state.
   */
  children: React.ReactElement<ButtonProps>[];
  /**
   * Sets the css class of the rendered element. Can be used to apply custom
   * styles.
   */
  className?: string;
}

const CompactButtonGroup = ({
  children,
  className,
}: CompactButtonGroupProps): React.ReactElement => {
  const childrenProps = React.Children.map(children, ({ props }) => props);

  if (!childrenProps.every(({ size }) => size === childrenProps[0].size)) {
    throw Error('All Buttons in CompactButtonGroup must be the same size');
  }

  if (
    !childrenProps.every(
      ({ appearance, intent, disabled, size }) =>
        appearance === childrenProps[0].appearance &&
        disabled === childrenProps[0].disabled &&
        intent === childrenProps[0].intent &&
        size === childrenProps[0].size,
    )
  ) {
    throw Error(
      'All Buttons in compact CompactButtonGroup must have the same appearance, intent, size & disabled state',
    );
  }

  const buttonCount = React.Children.count(children);
  const radii = { 0: '4px 0px 0px 4px', [buttonCount - 1]: '0px 4px 4px 0px' };
  const isPrimaryButtonGroup = childrenProps[0].appearance === 'primary';
  return (
    <div
      className={className}
      css={{ display: 'inline-flex', whiteSpace: 'nowrap' }}
    >
      <ClassNames>
        {({ css }) =>
          React.Children.map(children, (child, index) => {
            return React.cloneElement(child, {
              className: css({
                ':active': { transform: 'none' },
                ':focus': { boxShadow: 'none' },
                borderLeft:
                  index > 0 && isPrimaryButtonGroup
                    ? '1px solid rgba(5, 25, 45, 0.3)'
                    : undefined,
                borderRadius: radii[index] || '0px', // only round corners on first & last child
                margin: 0,
                marginLeft: index > 0 ? '-2px' : undefined, // so only 1 border shows between buttons
              }),
            });
          })
        }
      </ClassNames>
    </div>
  );
};

CompactButtonGroup.propTypes = {
  children: PropTypes.arrayOf(childrenOfType(Button).isRequired).isRequired,
} as React.WeakValidationMap<CompactButtonGroupProps>;

export default CompactButtonGroup;
