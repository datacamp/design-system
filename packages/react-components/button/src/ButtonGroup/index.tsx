import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import { childrenOfType, isChildType } from '@datacamp/waffles-utils';
import { ClassNames } from '@emotion/react';
import PropTypes from 'prop-types';
import React from 'react';

import Button from '../Button';
import CompactButtonGroup from '../CompactButtonGroup';
import {
  FacebookButton,
  GoogleButton,
  LinkedinButton,
  TwitterButton,
} from '../SocialMediaButtons';

interface ButtonGroupProps {
  children: React.ReactNode;
  /**
   * Sets the css class of the rendered element. Can be used to apply custom
   * styles.
   */
  className?: string;
}

const ButtonGroup = ({
  children,
  className,
}: ButtonGroupProps): React.ReactElement => {
  const sizes =
    React.Children.map(children, (child) => {
      if (!React.isValidElement(child)) return undefined;
      if (isChildType(child, Button)) return child.props.size || 'medium';
      if (isChildType(child, CompactButtonGroup)) {
        return child.props.children[0].props.size || 'medium'; // we already validate that all buttons in child group are the same size
      }
      return undefined;
    }) || [];

  if (!sizes.every((size) => size === sizes[0])) {
    throw Error('All Buttons in ButtonGroup must be the same size');
  }

  const gap =
    sizes[0] === 'small'
      ? tokens.size.space[8].value
      : tokens.size.space[16].value;

  return (
    <div
      className={className}
      css={{ display: 'inline-flex', whiteSpace: 'nowrap' }}
    >
      <ClassNames>
        {({ css, cx }) =>
          React.Children.map(children, (child, index) => {
            if (child === null) return null;
            const childElement = child as React.ReactElement<{
              className: string;
            }>;
            return React.cloneElement(childElement, {
              className: cx(
                css({ marginLeft: index > 0 ? gap : 0 }),
                childElement.props.className,
              ),
            });
          })
        }
      </ClassNames>
    </div>
  );
};

ButtonGroup.propTypes = {
  /**
   * The content of the ButtonGroup to render. This must be a set of Button
   * and/or CompactButtonGroup components. All Buttons within a ButtonGroup must
   * be the same size.
   */
  children: PropTypes.arrayOf(
    PropTypes.oneOfType([
      childrenOfType(Button),
      childrenOfType(FacebookButton),
      childrenOfType(GoogleButton),
      childrenOfType(LinkedinButton),
      childrenOfType(TwitterButton),
      childrenOfType(CompactButtonGroup),
    ]),
  ).isRequired,
};

export default ButtonGroup;
