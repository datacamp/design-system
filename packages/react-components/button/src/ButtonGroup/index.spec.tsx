import '@testing-library/jest-dom/extend-expect';

import axeRender from '@datacamp/waffles-axe-render';
import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import React from 'react';
import ErrorBoundary from 'react-error-boundary';

import Button from '../Button';
import CompactButtonGroup from '../CompactButtonGroup';
import ButtonGroup from '.';

describe('ButtonGroup', () => {
  it('renders the buttons or CompactButtonGroups with spaces', async () => {
    const { container, getByText } = await axeRender(
      <ButtonGroup>
        <Button type="submit">Button 1</Button>
        <Button type="submit">Button 2</Button>
        <Button type="submit">Button 3</Button>
        <CompactButtonGroup className="compact">
          <Button type="submit">Button 4</Button>
          <Button type="submit">Button 5</Button>
        </CompactButtonGroup>
        {false /* ensures conditional rendering is ok */}
      </ButtonGroup>
    );

    expect(
      (getByText('Button 1') as HTMLElement).closest('button')
    ).toHaveStyle('margin-left: 0;');

    expect(
      (getByText('Button 2') as HTMLElement).closest('button')
    ).toHaveStyle(`margin-left: ${tokens.size.space[16].value}px;`);

    expect(
      (getByText('Button 3') as HTMLElement).closest('button')
    ).toHaveStyle(`margin-left: ${tokens.size.space[16].value}px;`);

    expect(container.querySelector(`.compact`) as HTMLElement).toHaveStyle(
      `margin-left: ${tokens.size.space[16].value}px;`
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('adds the className to the group', async () => {
    const testClass = 'test';
    const { container } = await axeRender(
      <ButtonGroup className={testClass}>
        <Button type="submit">Button 1</Button>
        <Button type="submit">Button 2</Button>
        <Button type="submit">Button 3</Button>
      </ButtonGroup>
    );

    expect(container.firstChild).toHaveClass(testClass);
  });

  it('throws an error when different size buttons are provided', async () => {
    const spy = jest.spyOn(console, 'error');
    spy.mockImplementation(() => {});
    const onError = jest.fn();
    await axeRender(
      <ErrorBoundary onError={onError}>
        <ButtonGroup>
          <Button type="submit">Button 1</Button>
          <Button type="submit">Button 2</Button>
          <Button size="small" type="submit">
            Button 3
          </Button>
        </ButtonGroup>
      </ErrorBoundary>
    );

    expect(onError).toHaveBeenCalledTimes(1);
    expect(onError).toHaveBeenCalledWith(
      Error('All Buttons in ButtonGroup must be the same size'),
      expect.anything()
    );
  });

  it('throws an error when different size buttons are provided within a nested CompactButtonGroup', async () => {
    const spy = jest.spyOn(console, 'error');
    spy.mockImplementation(() => {});
    const onError = jest.fn();
    await axeRender(
      <ErrorBoundary onError={onError}>
        <ButtonGroup>
          <Button type="submit">Button 1</Button>
          <Button type="submit">Button 2</Button>
          <CompactButtonGroup>
            <Button size="small" type="submit">
              Button 3
            </Button>
            <Button size="small" type="submit">
              Button 4
            </Button>
          </CompactButtonGroup>
        </ButtonGroup>
      </ErrorBoundary>
    );

    expect(onError).toHaveBeenCalledTimes(1);
    expect(onError).toHaveBeenCalledWith(
      Error('All Buttons in ButtonGroup must be the same size'),
      expect.anything()
    );
  });
});
