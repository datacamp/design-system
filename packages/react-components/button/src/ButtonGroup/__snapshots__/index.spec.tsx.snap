// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup renders the buttons or CompactButtonGroups with 16px spaces 1`] = `
.emotion-11 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  white-space: nowrap;
}

.emotion-1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: capitalize;
  -webkit-transition: 0.15s;
  transition: 0.15s;
  vertical-align: baseline;
  white-space: nowrap;
  background-color: transparent;
  border-color: #33aacc;
  color: #33aacc;
  padding: 0 15px;
  margin-left: 0;
}

.emotion-1:active {
  -webkit-transform: perspective(1px) scale(0.975);
  -ms-transform: perspective(1px) scale(0.975);
  transform: perspective(1px) scale(0.975);
}

.emotion-1:disabled,
.emotion-1:hover:disabled,
.emotion-1:active:disabled {
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
}

.emotion-1:focus {
  outline: 0;
}

.emotion-1:focus {
  box-shadow: 0 0 0 4px rgba(51,170,204,0.2);
}

.emotion-1:hover {
  border-color: #2E8FAB;
  color: #2E8FAB;
}

.emotion-0 {
  -webkit-font-smoothing: antialiased;
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Lato;
  font-style: normal;
  font-size: 16px;
  font-weight: 400;
  line-height: 16px;
  font-size: 16px;
  line-height: 46px;
  color: #3d4251;
  font-weight: bold;
}

.emotion-3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: capitalize;
  -webkit-transition: 0.15s;
  transition: 0.15s;
  vertical-align: baseline;
  white-space: nowrap;
  background-color: transparent;
  border-color: #33aacc;
  color: #33aacc;
  padding: 0 15px;
  margin-left: 16px;
}

.emotion-3:active {
  -webkit-transform: perspective(1px) scale(0.975);
  -ms-transform: perspective(1px) scale(0.975);
  transform: perspective(1px) scale(0.975);
}

.emotion-3:disabled,
.emotion-3:hover:disabled,
.emotion-3:active:disabled {
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
}

.emotion-3:focus {
  outline: 0;
}

.emotion-3:focus {
  box-shadow: 0 0 0 4px rgba(51,170,204,0.2);
}

.emotion-3:hover {
  border-color: #2E8FAB;
  color: #2E8FAB;
}

.emotion-10 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  white-space: nowrap;
  margin-left: 16px;
}

.emotion-7 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: capitalize;
  -webkit-transition: 0.15s;
  transition: 0.15s;
  vertical-align: baseline;
  white-space: nowrap;
  background-color: transparent;
  border-color: #33aacc;
  color: #33aacc;
  padding: 0 15px;
  border-radius: 4px 0px 0px 4px;
  margin: 0;
}

.emotion-7:active {
  -webkit-transform: perspective(1px) scale(0.975);
  -ms-transform: perspective(1px) scale(0.975);
  transform: perspective(1px) scale(0.975);
}

.emotion-7:disabled,
.emotion-7:hover:disabled,
.emotion-7:active:disabled {
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
}

.emotion-7:focus {
  outline: 0;
}

.emotion-7:focus {
  box-shadow: 0 0 0 4px rgba(51,170,204,0.2);
}

.emotion-7:hover {
  border-color: #2E8FAB;
  color: #2E8FAB;
}

.emotion-7:active {
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
}

.emotion-7:focus {
  box-shadow: none;
}

.emotion-7:hover {
  z-index: 1;
}

.emotion-9 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: capitalize;
  -webkit-transition: 0.15s;
  transition: 0.15s;
  vertical-align: baseline;
  white-space: nowrap;
  background-color: transparent;
  border-color: #33aacc;
  color: #33aacc;
  padding: 0 15px;
  border-radius: 0px 4px 4px 0px;
  margin: 0;
  margin-left: -1px;
}

.emotion-9:active {
  -webkit-transform: perspective(1px) scale(0.975);
  -ms-transform: perspective(1px) scale(0.975);
  transform: perspective(1px) scale(0.975);
}

.emotion-9:disabled,
.emotion-9:hover:disabled,
.emotion-9:active:disabled {
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
}

.emotion-9:focus {
  outline: 0;
}

.emotion-9:focus {
  box-shadow: 0 0 0 4px rgba(51,170,204,0.2);
}

.emotion-9:hover {
  border-color: #2E8FAB;
  color: #2E8FAB;
}

.emotion-9:active {
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
}

.emotion-9:focus {
  box-shadow: none;
}

.emotion-9:hover {
  z-index: 1;
}

<div
  class="emotion-11"
>
  <button
    class="emotion-1"
    type="submit"
  >
    <span
      class="emotion-0"
    >
      Button 1
    </span>
  </button>
  <button
    class="emotion-3"
    type="submit"
  >
    <span
      class="emotion-0"
    >
      Button 2
    </span>
  </button>
  <button
    class="emotion-3"
    type="submit"
  >
    <span
      class="emotion-0"
    >
      Button 3
    </span>
  </button>
  <div
    class="compact emotion-10"
  >
    <button
      class="emotion-7"
      type="submit"
    >
      <span
        class="emotion-0"
      >
        Button 4
      </span>
    </button>
    <button
      class="emotion-9"
      type="submit"
    >
      <span
        class="emotion-0"
      >
        Button 5
      </span>
    </button>
  </div>
</div>
`;

exports[`ButtonGroup renders the small buttons or CompactButtonGroups with 8px spaces 1`] = `
.emotion-11 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  white-space: nowrap;
}

.emotion-1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: capitalize;
  -webkit-transition: 0.15s;
  transition: 0.15s;
  vertical-align: baseline;
  white-space: nowrap;
  background-color: transparent;
  border-color: #33aacc;
  color: #33aacc;
  padding: 0 15px;
  margin-left: 0;
}

.emotion-1:active {
  -webkit-transform: perspective(1px) scale(0.975);
  -ms-transform: perspective(1px) scale(0.975);
  transform: perspective(1px) scale(0.975);
}

.emotion-1:disabled,
.emotion-1:hover:disabled,
.emotion-1:active:disabled {
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
}

.emotion-1:focus {
  outline: 0;
}

.emotion-1:focus {
  box-shadow: 0 0 0 4px rgba(51,170,204,0.2);
}

.emotion-1:hover {
  border-color: #2E8FAB;
  color: #2E8FAB;
}

.emotion-7 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: capitalize;
  -webkit-transition: 0.15s;
  transition: 0.15s;
  vertical-align: baseline;
  white-space: nowrap;
  background-color: transparent;
  border-color: #33aacc;
  color: #33aacc;
  padding: 0 15px;
  border-radius: 4px 0px 0px 4px;
  margin: 0;
}

.emotion-7:active {
  -webkit-transform: perspective(1px) scale(0.975);
  -ms-transform: perspective(1px) scale(0.975);
  transform: perspective(1px) scale(0.975);
}

.emotion-7:disabled,
.emotion-7:hover:disabled,
.emotion-7:active:disabled {
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
}

.emotion-7:focus {
  outline: 0;
}

.emotion-7:focus {
  box-shadow: 0 0 0 4px rgba(51,170,204,0.2);
}

.emotion-7:hover {
  border-color: #2E8FAB;
  color: #2E8FAB;
}

.emotion-7:active {
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
}

.emotion-7:focus {
  box-shadow: none;
}

.emotion-7:hover {
  z-index: 1;
}

.emotion-9 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: capitalize;
  -webkit-transition: 0.15s;
  transition: 0.15s;
  vertical-align: baseline;
  white-space: nowrap;
  background-color: transparent;
  border-color: #33aacc;
  color: #33aacc;
  padding: 0 15px;
  border-radius: 0px 4px 4px 0px;
  margin: 0;
  margin-left: -1px;
}

.emotion-9:active {
  -webkit-transform: perspective(1px) scale(0.975);
  -ms-transform: perspective(1px) scale(0.975);
  transform: perspective(1px) scale(0.975);
}

.emotion-9:disabled,
.emotion-9:hover:disabled,
.emotion-9:active:disabled {
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
}

.emotion-9:focus {
  outline: 0;
}

.emotion-9:focus {
  box-shadow: 0 0 0 4px rgba(51,170,204,0.2);
}

.emotion-9:hover {
  border-color: #2E8FAB;
  color: #2E8FAB;
}

.emotion-9:active {
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
}

.emotion-9:focus {
  box-shadow: none;
}

.emotion-9:hover {
  z-index: 1;
}

.emotion-0 {
  -webkit-font-smoothing: antialiased;
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Lato;
  font-style: normal;
  font-size: 16px;
  font-weight: 400;
  line-height: 16px;
  font-size: 16px;
  line-height: 34px;
  color: #3d4251;
  font-weight: bold;
}

.emotion-3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: capitalize;
  -webkit-transition: 0.15s;
  transition: 0.15s;
  vertical-align: baseline;
  white-space: nowrap;
  background-color: transparent;
  border-color: #33aacc;
  color: #33aacc;
  padding: 0 15px;
  margin-left: 8px;
}

.emotion-3:active {
  -webkit-transform: perspective(1px) scale(0.975);
  -ms-transform: perspective(1px) scale(0.975);
  transform: perspective(1px) scale(0.975);
}

.emotion-3:disabled,
.emotion-3:hover:disabled,
.emotion-3:active:disabled {
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
}

.emotion-3:focus {
  outline: 0;
}

.emotion-3:focus {
  box-shadow: 0 0 0 4px rgba(51,170,204,0.2);
}

.emotion-3:hover {
  border-color: #2E8FAB;
  color: #2E8FAB;
}

.emotion-10 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  white-space: nowrap;
  margin-left: 8px;
}

<div
  class="emotion-11"
>
  <button
    class="emotion-1"
    type="submit"
  >
    <span
      class="emotion-0"
    >
      Button 1
    </span>
  </button>
  <button
    class="emotion-3"
    type="submit"
  >
    <span
      class="emotion-0"
    >
      Button 2
    </span>
  </button>
  <button
    class="emotion-3"
    type="submit"
  >
    <span
      class="emotion-0"
    >
      Button 3
    </span>
  </button>
  <div
    class="compact emotion-10"
  >
    <button
      class="emotion-7"
      type="submit"
    >
      <span
        class="emotion-0"
      >
        Button 4
      </span>
    </button>
    <button
      class="emotion-9"
      type="submit"
    >
      <span
        class="emotion-0"
      >
        Button 5
      </span>
    </button>
  </div>
</div>
`;
